### PR TITLE
Hz/msg loop

### PR DESCRIPTION
**Pull request by zhrt123**
bgworker:
- Disable refresh quota, instead add a demo function disk_quota_refresh().

center worker: A "static" bgworker created in _PG_init(), like launcher.
- Works as the server side in the message loop. Receives the request message,
  calls the corresponding function by message ID, and generates the response
  message.

message loop: A shared memory structure.
- server behavior
	- server always waits for `slatch`, the latch on the server side.
	- as soon as `slatch` is set, server gets a request message from client. 
	- handles the request.
	-  write the response message by `rsp_handle`.
	- set `clatch` to notify client to handle the response message.
- client behavior
	- waits for locking the `loop_lock`.
	- writes the request message by `req_handle`.
	- sets `slatch` to notify server to handle the request message.
	- waits for server to set `clatch`.
	- gets the response message after `clatch` is set.

TODO:
- How to handle Ctrl+C in UDF? (One solution is add pqsignal in each UDF)
- Should we destroy gangs after each iteration?
- Do we need disk_quota_sigusr1() on bgworker?
- Should we check the version for each bgworker?
- we should refactor or remove the `is_ready` judgment since it is too hard to read.
	- If we remove `diskquota.table_size`, the judgment could be removed.
	- Otherwise, we should refactor it.

### CODE REVIEW

File: src/msg_looper.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+typedef DiskquotaMessage *(*message_handler)(DiskquotaMessage *req_msg);
+typedef void (*signal_handler)(void);
+
+/* Called by server */
+/*
+ * To initialize a looper
+ * 1. request locks in the _pg_init() on the postmaster process
+ * request_message_looper_lock("my_looper");
+ * 2. create the looper struct on the server process (cannot be the postmaster process)
+ * looper = create_message_looper("my_looper");
```
Comment (beeender) [inline src/msg_looper.h:58]: Don't forget to mention to use `message_looper_size` to request shm.

File: src/msg_looper.c [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+/*
+ * Create a message.
+ */
+DiskquotaMessage *
+init_message(DiskquotaMessageID msg_id, size_t payload_len)
+{
+#if GP_VERSION_NUM < 70000
+	dsm_segment      *seg = dsm_create(MSG_SIZE(payload_len));
+#else
+	dsm_segment      *seg = dsm_create(MSG_SIZE(payload_len), 0);
```
[the remark this answers is not in the document]
Comment (zhrt123) [reply src/msg_looper.c]: We set the flag to 0 in gpdb7. And if there is no free dynamic shared memory segment, the error will be reported in gpdb.

File: src/msg_looper.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	int b;
+} TestMessage;
+
+typedef enum
+{
+	MSG_DEBUG = 10000,
+	MSG_TestMessage,
+	TIMEOUT_EVENT,
+} DiskquotaMessageID;
+
```
[the remark this answers is not in the document]
Comment (zhrt123) [reply src/msg_looper.h:25]: Added in c0ab82dc952d61a24a383f1c757e3672d78fefbf

File: src/msg_looper.c [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+/*------------------------------client---------------------------------*/
+/*
+ * Attach to the message looper.
+ * If the looper is not initialized yet, the return value is NULL.
+ */
+DiskquotaLooper *
+attach_message_looper(const char *name)
+{
+	bool found = false;
+	// FIXME: When to de-init?
```
Comment (beeender) [inline src/msg_looper.c:236]: create a `detach_message_looper()` function, client is responsible to detach it.

File: src/msg_looper.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+/*
+ * The message header.
+ * The message content follows the message header.,
+ * we can call MSG_BODY() to get the address of the message content.
+ */
+typedef struct DiskquotaMessage
```
[the remark this answers is not in the document]
Comment (beeender) [reply src/msg_looper.h:31]: Resolved? Shall we do this later (in a future PR)?

File: src/diskquota_center_worker.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	                  0);
+#else
+	BackgroundWorkerInitializeConnection(DISKQUOTA_DB, NULL, 0);
+	set_config_option("application_name", DISKQUOTA_APPLICATION_NAME, PGC_USERSET, PGC_S_SESSION, GUC_ACTION_SAVE, true,
+	                  0, true);
+#endif /* GP_VERSION_NUM */
+
+	init_ps_display("center worker:", "[diskquota]", DISKQUOTA_DB, "");
+
+	CurrentResourceOwner    = ResourceOwnerCreate(NULL, DISKQUOTA_CENTER_WORKER_NAME);
```
[the remark this answers is not in the document]
Comment (zhrt123) [reply src/diskquota_center_worker.c:123]: We need to create resource owner for dsm_create

File: src/diskquota_center_worker.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	DiskquotaLooper *looper = create_message_looper(DISKQUOTA_CENTER_WORKER_MESSAGE_LOOPER_NAME);
+	Assert(looper != NULL);
+	init_message_looper(looper, disk_quota_message_handler);
+
+	// FIXME: should we destroy gangs?
```
[the remark this answers is not in the document]
Comment (zhrt123) [reply src/diskquota_center_worker.c:128]: hmmm, we can do that.

File: src/msg_looper.c [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+	 *        loop.
+	 * - if request_done is true, then the request message is set completely.
+	 */
+	looper->request_done = true;
+	SetLatch(looper->slatch);
+
+	do
+	{
+		looper_wait_for_client_latch(looper);
+		if (handler != NULL) handler();
```
[the remark this answers is not in the document]
Comment (zhrt123) [reply src/msg_looper.c:302]: If the sigterm is received, the process should exit.

File: src/msg_looper.c (hunk 4 above)
Comment (zhrt123) [reply src/msg_looper.c:236]: hmm, what does the client need to do in detach function?

**Comment by zhrt123** [comment]
> Resolved? Shall we do this later (in a future PR)?

Miss operation. Let's do this in the next PR.

File: src/msg_looper.c [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+void
+message_looper_handle_message(DiskquotaLooper *looper)
+{
+	/* handle the message sent by bgworker */
+	if (looper->request_done || looper->req_handle != DSM_HANDLE_INVALID)
+	{
+		DiskquotaMessage *req_msg = attach_message(looper->req_handle);
+
+		/* free the response message generated by the last connection. */
+		free_message_by_handle(looper->rsp_handle);
```
[the remark this answers is not in the document]
Comment (zhrt123) [reply src/msg_looper.c:213]: The 5th step can be avoided by setting looper->req_handle to DSM_HANDLE_INVALID before setting clatch in the server.

File: src/diskquota_center_worker.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (beeender) [reply src/diskquota_center_worker.c:123]: Add a comment to say why we do NOT have to revert back to the old resource owner.

File: src/msg_looper.c (hunk 4 above)
Comment (beeender) [reply src/msg_looper.c:247]: OK. No need to de-init the shmem. This is OK.

**Review by beeender — APPROVED** [review]

**Pull request merged**